### ISSUE TITLE
speed up video sending when compressing the image

### ISSUE THREAD

**Issue by julio16101** [abhiTronix/vidgear#65]
<! --- Provide a general summary of the problem in the previous Title ->
Accelerate video streaming by compressing information
## Detailed description
<! --- Provide a detailed description of the change or addition you are proposing ->
NetGear works correctly for sending video between clients and servers, but it can improve this speed in situations where high quality video is handled or in saturated networks, reducing the size when compressing the viden sent.

What I propose is, internally add a preprocessing of the image, compressing it in jpeg format, with a variable that indicates to what quality it is correct to compress it, it could even be compressed as png.

Something like, before sending it, apply the compression:

```
encode_param = [int (cv2.IMWRITE_JPEG_QUALITY), 90]
result, encimg = cv2.imencode ('. jpg', img, encode_param)
```

and upon receiving it, decompress it:

`decimg = cv2.imdecode (encimg, 1)`

information here:
https://docs.opencv.org/master/d4/da8/group__imgcodecs.html#ga292d81be8d76901bff7988d18d2b42ac

### Context
<! --- Why is this change important to you? How would you use it? ->
I think it would be helpful to transmit video in large modifications where its weight in memory causes a slow delivery. In addition to processes where the image quality is not of great importance.
<! --- Will this change the existing VidGear APIs? How? ->
It would not affect much, due to serious compression and decompression from point to point, when sending and receiving it.
<! --- How can it benefit other users? ->
Helping them to speed up the process of sending information.

### Its environment
<! --- Include so many relevant details about the environment in which you worked ->
* VidGear version: v0.1.5
* Branch: <! --- Teacher / Testing / Development / PyPi -> Teacher
* Python version: 3.5
* pip version: 19.3.1
* Operating system and version: Linux

### Any other important information
<! --- This is an example / screenshot that I want to share ->
![mejora](https://user-images.githubusercontent.com/37164846/68159182-e7830e80-ff16-11e9-8c97-d75fc8bb4dd4.jpg)

**Comment by abhiTronix** [comment]
**Hello@julio16101 Thanks for this but I've some concerns:**

> What I propose is, internally add a preprocessing of the image, compressing it in jpeg format, with a variable that indicates to what quality it is correct to compress it, it could even be compressed as png.
> Something like, before sending it, apply the compression:
> encode_param = [int (cv2.IMWRITE_JPEG_QUALITY), 90]
> result, encimg = cv2.imencode ('. jpg', img, encode_param)

**Don't we have to save every video frame to a compressed png/jpeg before transmitting ?, if yes, then it will take significant amount of time and memory on your disk, which make it a Non-Optimal solution.**

### Solution:
**But instead I purpose to use cv2::resize() for compression like `cv2.resize(src, Size(), factor, factor, interpolation)` here, interpolation can be selected as `cv2.INTER_LANCZOS4` to obtain best interpolation results. OR we can use more stable Pillow resize instead for this purpose.** 


### Related Info:

_Someone had proposed a [issue related to stream compression in ZMQ](https://github.com/zeromq/libzmq/issues/2568), but it isn't been implemented yet._

**Comment by julio16101** [comment]
In this case, it would only be to temporarily compress the video image to be sent, in a variable, not necessarily store it on disk.

Only the frame to send would be in the output track, a variable could handle this without the need to store it on disk.

I have already done this compression in jpeg and png using imagezmq of jeffbass and the result is optimal.

I can conclude that although there may be a small waiting time when performing the compression process, the gain time is much longer.

In my experiments that I have done using imagezmq, I have obtained 50% more speed, even being able to double the speed but with the fact of dispensing with image quality, although in this case the compression could also be applied to png which also gives good results.

On the weekend I tried to apply the compression in his library, I was able to send the video frame by converting the compressed tuple to a numpy value, but upon receiving the video frame to apply the decompression I was not able to recover it correctly.

Your idea of ​​applying resize is good, but we would be subject to that resolution, where if we wanted to restore it back to its original resolution, we could lose more quality than applying jpeg compression. Although compression in PNG is more optimal.

**Comment by abhiTronix** [comment]
**@julio16101 Thanks for the detailed insight. After testing this algorithm on my machine, I'll roll out PR related to this enhancement in upcoming days. Kindly stay tuned and ready to test.**

**Comment by abhiTronix** [comment]
@julio16101 According to an answer:https://answers.opencv.org/question/207286/why-imencode-taking-so-long/?answer=211496#post-id-211496

The time varies for encoding in different format as follows:
**Time Calculations  while encoding:**
(A) Encoding with bmp : 20-40ms
(B) Encoding with jpg : 50-70 ms
(C) Encoding with png: 200-250ms 

Despite `bmp` is fasted, I think using `jpg` will be suitable  for encoding since optimized turbojpeg is now a part of OpenCV  binaries. But I'll gave user the choice to choose the best for them.

**Comment by abhiTronix** [comment]
@julio16101 , Thanks for your patience, **Kindly install the PR(#72 ) as follows:**

```sh
git clone https://github.com/abhiTronix/vidgear.git
cd vidgear
git checkout development
sudo pip3 install .
cd
```

**and test this example: https://github.com/abhiTronix/vidgear/wiki/Compression-in-NetGear-API#a-server-endbare-minimum-example and if it works as you expected or not? and then revert here. Good luck!**

**Comment by julio16101** [comment]
I will verify the code these days.

But first I will do a clean installation on the raspberry and opencv.

I will notify as soon as I try

**Comment by julio16101** [comment]
Sorry for the delay.

I've tried the new function, it's just as I expected. The transmission speeds of can increase to a high degree, of course although losing some quality.

I've tried jpeg and png so far.
In jpeg I recommend using "IMWRITE_JPEG_QUALITY" with values between 70-80 to further increase speed and maintain quality to some degree.

Although "IMWRITE_JPEG_OPTIMIZE" has the best quality-speed relationship.

Try the compression in png, where my results were not so good, opting for the minimum compression value in png, my speed was reduced to less than 1 FPS. It may be likely due to the power of the raspberry pi.

Keep playing with the values.
